web/src/components/map/map.jsx

```diff
@@ -0,0 +1,65 @@
+import React, { useEffect } from 'react';
+import PropTypes from 'prop-types';
+
+import { getBlockIndex, addDefaultColumn } from './utils';
+import {
+  BLOCKS_ACROSS,
+  BLOCKS_DOWN,
+  BLOCK_SIZE,
+  MOVING_SPEED,
+} from './constants';
+import styles from './map.module.css';
+
+export default function Map({ map, updateMap, time }) {
+  useEffect(() => {
+    const canvas = document.getElementById('map-canvas');
+    const context = canvas.getContext('2d');
+    context.clearRect(0, 0, canvas.width, canvas.height);
+
+    function fillBlock(colStart, rowStart, color) {
+      context.fillStyle = color;
+      context.fillRect(colStart, rowStart, BLOCK_SIZE, BLOCK_SIZE);
+
+      context.beginPath();
+      context.rect(colStart, rowStart, BLOCK_SIZE, BLOCK_SIZE);
+      context.stroke();
+    }
+
+    const adjustedX = Math.floor(time / MOVING_SPEED) % BLOCK_SIZE;
+
+    for (let row = 0; row < canvas.height; row += BLOCK_SIZE) {
+      const pixelRow = Math.floor(row / BLOCK_SIZE);
+      for (let col = 0 - adjustedX; col < canvas.width; col += BLOCK_SIZE) {
+        const pixelCol = Math.floor(col / BLOCK_SIZE) + 1;
+        fillBlock(col, row, map[getBlockIndex(pixelRow, pixelCol)]);
+      }
+    }
+
+    // if entering brand new col
+    if (adjustedX === 0) {
```

Comment:
afaict this is not guaranteed to run every time - if a computer is under some graphical load (e.g. from another program running) it's entirely possible that the time between requestAnimationFrames is such that adjustedX jumps from `BLOCK_SIZE - 1` (or less) to `1` (or greater) without passing 0 (since it’s a function of the time)